### ISSUE TITLE
Allow custom error handling depending on the HTTP status code

### ISSUE THREAD

**Issue by dachev** [square/picasso#2124]
It would be nice to be able to handle network failures depending on the HTTP status code. 

Currently, when image downloading fails, Picasso executes Callback's onError(Exception e) method and passes a NetworkRequestHandler.ResponseException. However, because this is a package-private class, we can't cast the generic argument to get the status code.

I think that throwing a specialized exception when the consumer can only "see" its generic interface defeats the purpose.

**Comment by JakeWharton** [comment]
Agreed. We should have a public HttpException similar to Retrofit which provides access to the response information.
